node/src/components/block_validator.rs

```diff
@@ -229,6 +254,45 @@ where
     }
 }
 
+/// Returns effects that fetch the deploy and validate it.
+fn fetch_deploy<REv, T, I>(
+    effect_builder: EffectBuilder<REv>,
+    chainspec: Arc<Chainspec>,
+    block_timestamp: Timestamp,
+    deploy_hash: DeployHash,
+    sender: I,
+) -> Effects<Event<T, I>>
+where
+    REv: From<Event<T, I>>
+        + From<BlockValidationRequest<T, I>>
+        + From<StorageRequest>
+        + From<FetcherRequest<I, Deploy>>
+        + Send,
+    T: BlockLike + Debug + Send + Clone + 'static,
+    I: Clone + Send + PartialEq + Eq + 'static,
+{
+    let validate_deploy = move |result: FetchResult<Deploy, I>| match result {
+        FetchResult::FromStorage(deploy) | FetchResult::FromPeer(deploy, _) => {
+            if deploy
+                .header()
+                .is_valid(&chainspec.deploy_config, block_timestamp)
+            {
+                Event::DeployFound(deploy_hash)
+            } else {
+                info!(%deploy_hash, "deploy is invalid");
+                Event::DeployMissing(deploy_hash, false)
```

Comment:
Isn't this "invalid" rather than missing?

Reply:
There's just one event variant `DeployMissing` for both invalid deploys and timed out requests. In this case, deploy failed the test against the chainspec so it is invalid. The boolean is interpreted as `timed_out` so `false` here means it's invalid.

Reply:
Sure, I'm just thinking it would be an improvement to have an "invalid" variant. I won't block the pr on that though.

Reply:
Done in [`37fd3f7` (#882)](https://github.com/CasperLabs/casper-node/pull/882/commits/37fd3f7ead882e9da4856cdb8e930d2a3444f674)